quil-py/quil/program/__init__.pyi

```diff
@@ -238,6 +314,73 @@ class CalibrationSet:
         """
         ...
 
+
+@final
+class FixedScheduleItem:
+    """
+    A single item within a fixed schedule, representing a single instruction within a basic block.
+    """
+    @property
+    def instruction_index(self) -> int:
+        """
+        The index of the instruction within the basic block.
+        """
+    @property
+    def time_span(self) -> FixedTimeSpan:
+        """
+        The time span during which the instruction is scheduled.
+        """
+
+
+@final
+class ControlFlowGraph:
+    """
+    Representation of a control flow graph (CFG) for a Quil program.
+
+    The CFG is a directed graph where each node is a basic block and each edge is a control flow
+    transition between two basic blocks.
+    """
+
+    def has_dynamic_control_flow(self) -> bool: 
+        """
+        Return ``True`` if the program has dynamic control flow, i.e. contains a conditional branch instruction.
+        """
+    def basic_blocks(self) -> List["BasicBlock"]: 
+        """
+        Return a list of all the basic blocks in the control flow graph, in order of definition.
+        """
+
+@final
+class FixedSchedule:
+    def items(self) -> List[FixedScheduleItem]:
+        """
+        All the items in the schedule, in unspecified order.
+        """
+    def duration(self) -> float:
+        """
+        The duration of the schedule, in seconds.
+
+        This is the maximum of the end time of all the items.
+        """
+
+@final
+class FixedTimeSpan:
```

Comment:
To save the user the math, I'd recommend adding an `end()` property documented as a function of the other two. However, I can also see this being more appropriate to the `pyquil` plumbing, which is yet to come.

Reply:
yep, I had considered it but left it off to keep the API minimal. Will add.

Reply:
Looks good.